swap/src/network/tor_transport.rs

```diff
@@ -0,0 +1,124 @@
+use data_encoding::BASE32;
+use futures::future::Ready;
+use futures::prelude::*;
+use libp2p::core::multiaddr::{Multiaddr, Protocol};
+use libp2p::core::transport::TransportError;
+use libp2p::core::Transport;
+use libp2p::tcp::tokio::{Tcp, TcpStream};
+use libp2p::tcp::{GenTcpConfig, TcpListenStream, TokioTcpConfig};
+use std::io;
+use std::net::{Ipv4Addr, SocketAddr, SocketAddrV4};
+use std::pin::Pin;
+use tokio_socks::tcp::Socks5Stream;
+use tokio_socks::IntoTargetAddr;
+
+/// Represents the configuration for a TCP/IP transport capability for libp2p.
+#[derive(Clone)]
+pub struct TorTcpConfig {
+    inner: GenTcpConfig<Tcp>,
+    /// Tor SOCKS5 proxy port number.
+    socks_port: u16,
+}
+
+impl TorTcpConfig {
+    pub fn new(tcp: TokioTcpConfig, socks_port: u16) -> Self {
+        Self {
+            inner: tcp,
+            socks_port,
+        }
+    }
+}
+
+impl Transport for TorTcpConfig {
+    type Output = TcpStream;
+    type Error = io::Error;
+    type Listener = TcpListenStream<Tcp>;
+    type ListenerUpgrade = Ready<Result<Self::Output, Self::Error>>;
+    #[allow(clippy::type_complexity)]
+    type Dial = Pin<Box<dyn Future<Output = Result<Self::Output, Self::Error>> + Send>>;
+
+    fn listen_on(self, addr: Multiaddr) -> Result<Self::Listener, TransportError<Self::Error>> {
+        self.inner.listen_on(addr)
+    }
+
+    // dials via Tor's socks5 proxy if configured and if the provided address is an
+    // onion address. or it falls back to Tcp dialling
+    fn dial(self, addr: Multiaddr) -> Result<Self::Dial, TransportError<Self::Error>> {
+        async fn do_tor_dial(socks_port: u16, dest: String) -> Result<TcpStream, io::Error> {
+            tracing::trace!("Connecting through Tor proxy to address: {}", dest);
+            let stream = connect_to_socks_proxy(dest, socks_port)
+                .await
+                .map_err(|e| io::Error::new(io::ErrorKind::ConnectionRefused, e))?;
+            tracing::trace!("Connection through Tor established");
+            Ok(stream)
+        }
+
+        match to_onion_address(addr.clone()) {
+            Some(tor_address_string) => {
+                Ok(Box::pin(do_tor_dial(self.socks_port, tor_address_string)))
+            }
+            _ => self.inner.dial(addr),
+        }
+    }
```

Comment:
@thomaseizinger/ @da-kami : I thought ab it more more on this particular part: Assuming the user configured a Socks5 proxy port we should assume that she wants all connections to go through that port, i.e. if the user provides a Tor socks5 port, we should always dial through this port.  

As of now, we only dial through the socks5 port if we received an onion address. 
Now, if we always want to do this, we need to manually translate the provided multiaddress into a string of format `ADDR:PORT`. I tried to achieve this on my playground project to send all traffic through Tor but failed so far. It could be that this is because Tor blocks traffic to localhost, so this would need some more investigation.

Reply:
> if the user provides a Tor socks5 port, we should always dial through this port.

I agree.